docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,189 @@
+
+## Sync Process
+The `LDAPMemberLister` inside of the `LDAPGroupSyncer` will, for each group that needs to be synced, find the list of entries in the LDAP server that encode the members of the group. The `LDAPCachedMemberLister` has the same functionality for LDAP sync jobs where it is not possible to run an LDAP query to get the members of each group to be synced: when syncing all groups for an LDAP schema where groups are attributes, *etc*. The `LDAPMemberIdentityFactory` creates OpenShift `Identities` for these members. The `UserIdentityMapper` determines the OpenShift `Users` that these `Identities` refer to. Syncing all or some of either LDAP or OpenShift groups is supported.
+
+```go
+type LDAPGroupSyncer struct {
+  options            LDAPSyncOptions
+  filterGenerator    LDAPFilterGenerator
+  memberLister       LDAPMemberLister
+  cachedMemberLister LDAPCachedMemberLister
+  identityFactory    LDAPMemberIdentityFactory
+  identityMapper     *userapi.UserIdentityMapper
+  client             *osclient.Client
+}
+
+func (s *LDAPGroupSyncer) SyncAllInternalGroups() (err error) {
+  // use s.client to determine Groups to sync
+  s.SyncInternalGroups(groups)
+}
+
+func (s *LDAPGroupSyncer) SyncInternalGroups(groups *[]Group) (err error) {
+  // use s.filterGenerator to map Group -> groupFilter
+  s.SyncGroups(groupFilters)
+}
+
+func (s *LDAPGroupSyncer) SyncAllGroups() (err error) {
+  // use s.cachedMemberLister to determine groups to sync and find members for each
+  // for each group:
+  //    users := s.determineUsers(members)
+  // use s.groupClient to determine which Group to update
+}
+
+func (s *LDAPGroupSyncer) SyncGroups(groupFilters util.StringSet) (err error) {
+  // use s.memberLister to map groupFilter -> members
+  // for each group:
+  //    users := s.determineUsers(members)
+  // use s.groupClient to determine which Group to update
+}
+
+func (s *LDAPGroupSyncer) determineUsers(members []*ldap.Entry) (users []*userapi.User, err error) {
+  // use s.identityFactory to map member -> Identity
+  // use s.identityMapper to determine Identity -> user
+}
+
+```
+`LDAPSyncOptions` needs the following schema (TBD) :
+```
+LDAP server information (would this be in the identityprovider?)
+LDAP schema has first class groups?
+common LDAP filter for "all" (groups, users)
+specific filter key (for groups and for users)
+mapping OpenShift Group to specific filter value (either by opaque values or programatically from Group fields)
+mapping LDAP group to specific filter value
+how to map LDAP group attribute specifying a member to specific filter value
+attribute on first-class member that specifies groups or on first-class group that specifies members
+mapping from first-class LDAP group attributes to OpenShift Group fields
+```
+
+##  Determining Which Groups To Sync
```

Comment:
I see these cases:
 1. from whitelist file (part of the config)?
 1. from ldap query minus a blacklist file (part of the config)?
 1. from specific list of openshift groups using the label.

I think that general label selection cases will be handled via `oc get -l selector -t template | xargs oc sync`